components/org.wso2.carbon.uis/src/main/java/org/wso2/carbon/uis/internal/io/deployment/ArtifactAppDeployer.java

```diff
@@ -27,125 +27,170 @@
-        stop();
+        try {
+            for (String artifactKey : deployedApps.keySet()) {
+                undeploy(artifactKey);
+            }
+        } catch (CarbonDeploymentException e) {
+            throw new UISRuntimeException("An error occurred during deactivation of app deployer.", e);
+        }
         LOGGER.debug("Carbon UI server app deployer deactivated.");
     }
 
-    private void start() {
-        if (!Files.exists(appsRepository)) {
-            LOGGER.debug("Web apps repository '{}' does not exists.", appsRepository.toString());
-            return;
-        }
+    @Override
+    public void init() {
+        LOGGER.debug("Carbon UI server app deployer initialized.");
+    }
 
-        Set<AppReference> appReferences;
-        try {
-            appReferences = Files.list(appsRepository)
-                    .filter(Files::isDirectory)
-                    .map(ArtifactAppReference::new)
-                    .collect(Collectors.toSet());
-        } catch (IOException e) {
-            throw new DeploymentException("Cannot list web apps in '" + appsRepository + "' directory.", e);
+    @Override
+    public Object deploy(Artifact artifact) throws CarbonDeploymentException {
+        Path appPath = artifact.getFile().toPath();
+        if (!isValidAppArtifact(appPath)) {
+            throw new DeploymentException("Artifact located in '" + appPath + "'is not a valid web app.");
         }
-        publishAppsDeploymentEvents(appReferences);
+
+        App app = createApp(appPath);
+        String artifactKey = "webapp:" + app.getName();
+
+        appDeploymentEventListener.appDeploymentEvent(app);
+        deployedApps.put(artifactKey, app);
+        LOGGER.debug("Web app '{}' deployed for context path '{}'.", app.getName(), app.getContextPath());
+        return artifactKey;
     }
 
-    private void stop() {
-        for (String appName : deployedAppNames) {
-            publishAppUndeploymentEvent(appName);
+    @Override
+    public void undeploy(Object key) throws CarbonDeploymentException {
+        App app = deployedApps.remove(key.toString());
+        if (app == null) {
+            throw new CarbonDeploymentException("Web app for key '{}' cannot be found to undeploy.");
         }
+
+        appDeploymentEventListener.appUndeploymentEvent(app.getName());
+        LOGGER.debug("Web app '{}' undeployed from context path '{}'.", app.getName(), app.getContextPath());
     }
 
-    private void publishAppDeploymentEvent(AppReference deployingAppReference) {
-        for (AppDeploymentEventListener appDeploymentEventListener : eventListeners) {
-            appDeploymentEventListener.appDeploymentEvent(deployingAppReference);
-        }
-        deployedAppNames.add(deployingAppReference.getName());
+    @Override
+    public Object update(Artifact artifact) throws CarbonDeploymentException {
+        // when an app artifact gets updated, we just re-deploy it
```

Comment:
Shall we add a debug log to mark that we received an update call?

Reply:
fixed in commit a1e18fbf9c090aabc2382ce7ce1655f456b3576f